src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -188,46 +186,106 @@ private void EnqueueProcessSnapshot(Document newDocument)
+
+                // Once we're past this point, we're mutating our internal state so we cannot cancel past that.
+                cancellationToken = CancellationToken.None;
+
                 lock (_gate)
                 {
-                    _lastProcessedSnapshot = snapshot;
-                    _lastProcessedDocument = document;
+                    _lastProcessedSnapshot = currentSnapshot;
+                    _lastProcessedDocument = currentDocument;
+                    _lastProcessedSyntaxTree = currentSyntaxTree;
                 }
 
                 _reportChangeCancellationSource = new CancellationTokenSource();
                 _notificationService.RegisterNotification(() =>
                     {
                         _workQueue.AssertIsForeground();
-                        ReportChangedSpan(snapshot.GetFullSpan());
+                        ReportChangedSpan(changedSpan);
                     },
                     ReportChangeDelayInMilliseconds,
                     _listener.BeginAsyncOperation("ReportEntireFileChanged"),
                     _reportChangeCancellationSource.Token);
             }
 
+            private async Task<SnapshotSpan> GetChangedSpanAsync(
+                Document previousDocument, Document currentDocument,
+                ITextSnapshot currentSnapshot, CancellationToken cancellationToken)
+            {
+                // Try to defer to the classification service to see if it can find a narrower range to recompute.
+                if (previousDocument != null)
+                {
+                    var classificationService = TryGetClassificationService(currentSnapshot);
+                    if (classificationService != null)
+                    {
+                        var changeRange = await ComputeChangeRangeAsync(
+                            previousDocument, currentDocument, classificationService, cancellationToken).ConfigureAwait(false);
+                        if (changeRange != null)
+                            return currentSnapshot.GetSpan(changeRange.Value.Span.Start, changeRange.Value.NewLength);
+                    }
+                }
+
+                // Couldn't compute a narrower range.  Just the mark the entire file as changed.
+                return currentSnapshot.GetFullSpan();
+            }
+
+            private static async Task<TextChangeRange?> ComputeChangeRangeAsync(
+                Document previousDocument, Document currentDocument,
+                IClassificationService classificationService, CancellationToken cancellationToken)
+            {
+                // We want to compute a minimal change, but we don't want this to run for too long.  So do the
+                // computation work in the threadpool, but also gate how much time we can spend here so that we can let
+                // the editor know about the size of the change asap.
+                using var linkedToken = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```

Comment:
i wasn't comfortable with the idea of us just spending arbitrary time computing the diff.  though it may not be an issue in practice as i imagine our diffing algorithm (later in this pr) should be blisteringly fast.